kernel/src/hil/gpio.rs

```diff
@@ -1,27 +1,65 @@
+/// Enum for configuring any pull-up or pull-down
+/// resistors on the GPIO pin.
 pub enum InputMode {
     PullUp,
     PullDown,
     PullNone,
 }
 
+/// Enum for selecting which edge to trigger interrupts
+/// on.
 pub enum InterruptMode {
-    Change,
     RisingEdge,
     FallingEdge,
+    EitherEdge,
 }
 
-pub trait GPIOPin {
-    fn enable_output(&self);
-    fn enable_input(&self, mode: InputMode);
+pub trait PinCtl {
+    /// Configure whether the pin should have a pull-up or pull-down resistor or
+    /// neither.
+    fn set_input_mode(&self, InputMode);
+}
+
+/// Interface for synchronous GPIO pins.
+pub trait Pin {
+    /// Configure the GPIO pin as an output pin.
+    fn make_output(&self);
+
+    /// Configure the GPIO pin as an input pin.
+    fn make_input(&self);
+
+    /// Disable the GPIO pin and put it into its lowest power
+    /// mode.
     fn disable(&self);
+
+    /// Set the GPIO pin high. It must be an output.
     fn set(&self);
+
+    /// Set the GPIO pin low. It must be an output.
     fn clear(&self);
+
+    /// Toggle the GPIO pin. It must be an output.
     fn toggle(&self);
+
+    /// Get the current state of an input GPIO pin.
     fn read(&self) -> bool;
+
+    /// Enable an interrupt on the GPIO pin. It must
+    /// be configured as an interrupt. The `identifier`
+    /// can be any value and will be returned to you
+    /// when the interrupt on this pin fires.
     fn enable_interrupt(&self, identifier: usize, mode: InterruptMode);
+
+    /// Disable the interrupt for the GPIO pin.
     fn disable_interrupt(&self);
 }
 
+/// Interface for users of synchronous GPIO. In order
+/// to receive interrupts, the user must implement
+/// this `Client` interface.
 pub trait Client {
```

Comment:
The more I play around with the time hils, and in particular the AlarmToTimer type, the more I disagree with this decision.

In the case of GPIO, it means that if you write a driver that uses a GPIO pin for interrupts (e.g., the TMP006), you have no way of ensuring that you're indeed set up to be the client of that pin. Instead, you have to rely on the platform integrator to remember to call `set_client`, in some chip-implementation specific way, at some point after initializing both your driver and the pin.